python/az/aro/azext_aro/_dynamic_validators.py

```diff
@@ -0,0 +1,286 @@
+from msrestazure.tools import parse_resource_id
+from msrestazure.azure_exceptions import CloudError
+from azext_aro._validators import validate_vnet, validate_cidr
+from azext_aro._rbac import has_role_assignment_on_resource
+from humanfriendly.terminal.spinners import AutomaticSpinner
+
+
+logger = get_logger(__name__)
+
+
+def can_do_action(perms, action):
+    for perm in perms:
+        for perm_action in perm.actions:
+            clean = re.escape(perm_action)
+            clean = re.match("(?i)^" + clean.replace(r"\*", ".*") + "$", action)
+            if clean:
+                return None
+        for not_action in perm.not_actions:
+            clean = re.escape(not_action)
+            clean = re.match("(?i)^" + clean.replace(r"\*", ".*") + "$", action)
+            if clean:
+                return f"{action} permission is disabled"
+    return f"{action} permission is missing"
+
+
+def validate_resource(client, key, resource, actions):
+    perms = client.permissions.list_for_resource(resource['resource_group'],
+                                                 resource['namespace'],
+                                                 "",
+                                                 resource['type'],
+                                                 resource['name'])
+
+    errors = []
+    for action in actions:
+        perms, perms_copy = tee(perms)
+        perms_list = list(perms_copy)
+        error = can_do_action(perms_list, action)
+        if error is not None:
+            row = [key, resource['name'], error]
+            errors.append(row)
+
+    return errors
+
+
+def get_subnet(client, subnet, subnet_parts):
+    try:
+        subnet_obj = client.subnets.get(subnet_parts['resource_group'],
+                                        subnet_parts['name'],
+                                        subnet_parts['child_name_1'])
+    except Exception as err:
+        if isinstance(err, ResourceNotFoundError):
+            raise InvalidArgumentValueError(
+                f"Invalid -- subnet, error when getting '{subnet}': {str(err)}") from err
+        raise CLIInternalError(
+            f"Unexpected error when getting subnet '{subnet}': {str(err)}") from err
+
+    return subnet_obj
+
+
+def dyn_validate_vnet(key):
+    def _validate_vnet(cmd, namespace):
+        errors = []
+        with AutomaticSpinner("Validating Virtual Network Permissions"):
+            vnet = getattr(namespace, key)
+
+            if not is_valid_resource_id(vnet):
+                raise RequiredArgumentMissingError(
+                    f"Must specify --vnet if --{key.replace('_', '-')} is not an id.")
+
+            validate_vnet(cmd, namespace)
+
+            parts = parse_resource_id(vnet)
+
+            network_client = get_mgmt_service_client(
+                cmd.cli_ctx, ResourceType.MGMT_NETWORK)
+
+            auth_client = get_mgmt_service_client(
+                cmd.cli_ctx, ResourceType.MGMT_AUTHORIZATION, api_version="2015-07-01")
```

Comment:
So the validation of the resourse_id at the top is slightly different for the subnet vs the vnet. The vnet will try and validate itself and fail if its not valid. Whereas the subnet will try and validate itself and if it fails it will try and validate the vnet and retrieve the subnet from the vnet so it wouldn't be a very nice refactor. However, I have refactored the retrieving of the parts, network_client and auth_client into a new function as this is identical between the functions